PEP 690: Lazy Imports (#2569)
python/peps: pep-0690.rst

each scope the imported object is used in, and to avoid the inefficiency
of repeated execution of the import opcodes at runtime. This means that
importing the main module of a program typically results in an immediate
cascade of imports of most or all of the modules that may ever be needed
by the program.

Consider the example of a Python command line program with a number of
subcommands. Each subcommand may perform different tasks, requiring the import
of different dependencies. But a given invocation of the program will only
execute a single subcommand, or possibly none (i.e. if just ``--help`` usage
info is requested). Top-level eager imports in such a program will result in
the import of many modules that will never be used at all; the time spent
(possibly compiling and) executing these modules is pure waste.

Some large Python CLIs, in an effort to improve startup time, make some imports
lazy by manually placing imports inline into functions to delay imports of
expensive subsystems. This manual approach is labor-intensive and fragile; one
misplaced import or refactor can easily undo painstaking optimization work.

> Existing import-hook-based solutions such as demandimport [1]_ are limited in

Reviewer: ```suggestion
Existing import-hook-based solutions such as `demandimport
<https://github.com/bwesterb/py-demandimport/>`_ are limited in
```

Use actual links here

that only certain styles of import can be made truly lazy (imports such as
``from foo import a, b`` will still eagerly import the module ``foo``) and they
impose additional runtime overhead on every module attribute access.

This PEP proposes a more comprehensive solution for lazy imports that does not
impose detectable overhead in real-world use. The implementation in this PEP

> has already demonstrated startup time wins up to 70% and memory-use wins up to
> 40% on real-world Python CLIs [2]_.

Reviewer: ```suggestion
has already `demonstrated
<https://github.com/facebookincubator/cinder/blob/cinder/3.8/CinderDoc/lazy_imports.rst>`_
startup time improvements up to 70% and memory-use reductions up to
40% on real-world Python CLIs.
```

As above, use inline links

deferred until the imported name is used. This could happen immediately (e.g.
on the very next line after the import statement) or much later (e.g. while
using the name inside a function being called by some other code at some later
time.)

For these top level imports, there are two exceptions which will make them
eager (not lazy): imports inside ``try``/``except``/``finally``  or ``with``
blocks, and star imports (``from foo import *``.) Imports inside
exception-handling blocks remain eager so that any exceptions arising from the
import can be handled. Star imports must remain eager since performing the
import is the only way to know which names should be added to the namespace.

Imports inside class definitions or inside functions/methods are not "top
level" and are never lazy.

Dynamic imports using ``__import__`` or ``importlib.import_module`` are also
never lazy.

Per-module opt out
------------------

Due to the backwards compatibility issues mentioned below, it may be necessary
to force some imports to be eager. In first-party code, this can be easily
accomplished via any module-level reference to the name, e.g. even re-assigning

> the name to itself will trigger the import:
> 
> .. code-block python

Reviewer: ```suggestion
the name to itself will trigger the import::

```

Fix the code block

    import foo
    
    # ensure 'foo' is eagerly imported
    foo = foo

> Another option that scales better to making multiple imports lazy is to place

Reviewer 2: ```suggestion
Another option that scales better to making multiple imports eager is to place
```

> them inside a ``try/finally``:
> 
> .. code-block py

Reviewer: ```suggestion
them inside a ``try/finally``::

```

    try:  # force these imports to be eager
        import foo
        import bar
    finally:
        pass

The more difficult case can occur if an import in third-party code that can't
easily be modified must be forced to be eager. For this purpose, we propose to
add an API to ``importlib`` that can be called early in the process to specify

> a list of module names within which all imports will be eager:
> 
> .. code-block py

Reviewer: ```suggestion
a list of module names within which all imports will be eager::

```

    from importlib import set_eager_imports

    set_eager_imports(["one.mod", "another"])

Backwards Compatibility
=======================

This proposal preserves full backwards compatibility when the feature is
disabled, which is the default.

When enabled, lazy imports could produce currently unexpected results and
behaviors in existing codebases. The problems that we may see when enabling
lazy imports in an existing codebase are related to:

Import Side Effects
-------------------

Import side effects that would otherwise be produced by the execution of
imported modules during the execution of import statements will be deferred at
least until the imported objects are used.

These import side effects may include:

* code executing any side-effecting logic during import;
* relying on imported submodules being set as attributes in the parent module.

Dynamic Paths
-------------

There could be issues related to dynamic Python import paths; particularly,

> adding (and then removing after the import) paths from ``sys.path``.
> 
> .. code-block py

Reviewer: ```suggestion
adding (and then removing after the import) paths from ``sys.path``::

```

  foo.bar; foo.bar.Baz``, not ``import foo; foo.bar.Baz``. The latter only
  works (unreliably) because the attribute ``foo.bar`` is added as a side
  effect of ``foo.bar`` being imported somewhere else. With lazy imports this
  may not always happen on time.

* Avoid using star imports, as those are always eager.

* When possible, do not import whole submodules. Import specific names instead;
  i.e.: do ``from foo.bar import Baz``, not ``import foo.bar`` and then
  ``foo.bar.Baz``. If you import submodules (such as ``foo.qux`` and
  ``foo.fred``), with lazy imports enabled, when you access the parent module's
  name (``foo`` in this case), that will trigger loading all of the sibling
  submodules of the parent module (``foo.bar``, ``foo.qux`` and ``foo.fred``),
  not only the one being accessed, because the parent module ``foo`` is the
  actual deferred object name.

* Don't use inline imports, unless absolutely necessary. Import cycles should
  no longer be a problem with lazy imports enabled, so there’s no need to add
  complexity or more opcodes in a potentially hot path.

Reference Implementation
========================

> The current reference implementation is available as part of Cinder [3]_.

Reviewer: ```suggestion
The current reference implementation is available as part of 
`Cinder <https://github.com/facebookincubator/cinder>`_.
```

(and in third-party dependencies) to use the lazy import syntax.

It would be possible to aim for a "shallow" laziness where only the top-level
imports of subsystems from the main module are made explicitly lazy, but then
imports within the subsystems are all eager. This is extremely fragile, though
-- it only takes one mis-placed import to undo the carefully constructed
shallow laziness. Globally enabling lazy imports, on the other hand, provides
in-depth robust laziness where you always pay only for the imports you use.

Half-lazy imports
-----------------

It would be possible to eagerly run the import loader to the point of finding
the module source, but then defer the actual execution of the module and
creation of the module object. The advantage of this would be that certain
classes of import errors (e.g. a simple typo in the module name) would be
caught eagerly instead of being deferred to the use of an imported name.

The disadvantage would be that the startup time benefits of lazy imports would
be significantly reduced, since unused imports would still require a filesystem
``stat()`` call, at least. It would also introduce a possibly non-obvious split

> between _which_ import errors are raised eagerly and which are delayed, when

Reviewer: ```suggestion
between *which* import errors are raised eagerly and which are delayed, when
```

I assume you want this to be italic?

lazy imports are enabled.

This idea is rejected for now on the basis that in practice, confusion about
import typos has not been an observed problem with the reference
implementation. Generally delayed imports are not delayed forever, and errors
show up soon enough to be caught and fixed (unless the import is truly unused.)

> References
> ==========
> 
> .. [1] demandimport (https://github.com/bwesterb/py-demandimport/)
> .. [2] Lazy Imports https://github.com/facebookincubator/cinder/blob/cinder/3.8/CinderDoc/lazy_imports.rst
> .. [3] Reference implementation (https://github.com/facebookincubator/cinder)
> 
> 

Reviewer: ```suggestion
```

No longer needed, as above

        import bar

Since imports within context managers are always eager, the ``eager_imports()``
context manager can just be an alias to a null context manager. The context
manager does not force all imports to be recursively eager: ``foo`` and ``bar``
will be imported eagerly, but imports within those modules will still follow
the usual laziness rules.

The more difficult case can occur if an import in third-party code that can't
easily be modified must be forced to be eager. For this purpose, we propose to
add an API to ``importlib`` that can be called early in the process to specify
a list of module names within which all imports will be eager::

    from importlib import set_eager_imports

    set_eager_imports(["one.mod", "another"])

The effect of this is also shallow: all imports within ``one.mod`` will be
eager, but not imports in all modules imported by ``one.mod``.

Backwards Compatibility
=======================

This proposal preserves full backwards compatibility when the feature is
disabled, which is the default.

Even when enabled, most code will continue to work normally without any
observable change (other than improved startup time and memory usage.)

> Namespace packages are not effected: they work just as they do currently,

Reviewer 3: ```suggestion
Namespace packages are not affected: they work just as they do currently,
```
